File: 146-lru-cache/lru-cache.py
class Node:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = self.prev = None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cash = {}
        self.left, self.right = Node(0, 0), Node(0, 0)
        self.left.next, self.right.prev = self.right, self.left

    def insert(self, node):
        prev, next = self.right.prev, self.right
        prev.next = next.prev = node
        node.next, node.prev = next, prev
    
    def remove(self, node):
        prev, next = node.prev, node.next
        prev.next, next.prev = next, prev

    def get(self, key: int) -> int:
        if key in self.cash:
            self.remove(self.cash[key])
            self.insert(self.cash[key])
            return self.cash[key].value
        else:
            return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cash:
            self.remove(self.cash[key])
        self.cash[key] = Node(key, value)
        self.insert(self.cash[key])
        if len(self.cash) > self.capacity:
            lru = self.left.next
            self.remove(lru)
            del self.cash[lru.key]


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 146-lru-cache/lru-cache.js
class Node {
    constructor(key, value){
        this.key = key
        this.value = value
        this.left = null
        this.right = null
    }
}

/**
 * @param {number} capacity
 */
var LRUCache = function(capacity) {
    this.cap = capacity
    this.cache = new Map()
    this.left = new Node(0, 0)
    this.right = new Node(0, 0)
    this.left.right = this.right
    this.right.left =this.left
};

LRUCache.prototype.remove = function(node){
    const left = node.left, right = node.right

    left.right = right
    right.left = left
}

LRUCache.prototype.add = function(node){
    const left = this.right.left, right = this.right

    left.right = node
    right.left = node
    node.left = left
    node.right = right
}

/** 
 * @param {number} key
 * @return {number}
 */
LRUCache.prototype.get = function(key) {
    if(this.cache.has(key)){
        const node = this.cache.get(key)
        this.remove(node)
        this.add(node)
        return node.value
    }
    return -1
};

/** 
 * @param {number} key 
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function(key, value) {
    if(this.cache.has(key)){
        this.remove(this.cache.get(key))
    }
    const newNode = new Node(key, value)
    this.cache.set(key, newNode)
    this.add(newNode)
    if(this.cache.size > this.cap){
        const lru = this.left.right
        this.remove(lru)
        this.cache.delete(lru.key)
    }
};

/** 
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */